### DISCUSSION TITLE
Batching type resolution

### DISCUSSION THREAD

**Discussion by hrkfdn** [ExpediaGroup/graphql-kotlin#1869] [General] [answered]
With #1514 the interface for type resolvers has changed to pass a single `representation` instead of multiple `representations`. This was done to be more streamlined with other subgraph libraries which makes sense.

However, it comes with the disadvantage that type resolution can not be batched anymore. In our case, we would do a batch load from DynamoDB when receiving `representations` as that is more efficient than running a query per requested type. With the current interface this is not possible.

What would be the way to reproduce such a logic? I have asked on Slack and a data loader was suggested. From my understanding that does not mean the data loader will automatically be integrated with the type resolver, so I'm guessing some other gluing logic is needed.

Maybe some documentation or helpers in `graphql-kotlin` could be helpful to make the migration to v7 a little easier?

**Comment by samuelAndalon** [comment]
Hello, thank you for reaching out, as you well mentioned, we decided to refactor the federated type resolution to match with other federation implementations,

you can refer to the [federation section](https://opensource.expediagroup.com/graphql-kotlin/docs/schema-generator/federation/apollo-federation) of the documentation to check how to use DataLoaders, the only constrain is that you will have to use the [FederatedTypePromiseResolver](https://opensource.expediagroup.com/graphql-kotlin/docs/schema-generator/federation/type-resolution/#provide-federatedtyperesolvers-to-federatedschema), `FederatedTypeSuspendResolver` is incompatible with DataLoader's technique as it [relies on coroutines
](https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/#dataloaders-and-coroutines)
[here you can learn more about how to use DataLoaders in graphql-kotlin](https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/)

**Comment by hrkfdn** [comment]
Hey @samuelAndalon, thanks for the response.

I'm not sure I fully understood. How would I use `FederatedTypePromiseResolver` with batch loading with a single `representation`?

**Comment by samuelAndalon** [comment, marked as answer]
>I'm not sure I fully understood. How would I use FederatedTypePromiseResolver with batch loading with a single representation?

1. Define your resolvers for federated entities
```kotlin
// This service does not own the "Product" type but is extending it with new fields
@KeyDirective(fields = FieldSet("id"))
@ExtendsDirective
class Product(@ExternalDirective val id: String) {
  fun newField(): String = getNewFieldByProductId(id)
}

// This is how the "Product" class is created from the "_entities" query using promise resolver
class ProductResolver : FederatedTypePromiseResolver<Product> {
    override val typeName: String = "Product"

    override fun resolve(
        environment: DataFetchingEnvironment,
        representation: Map<String, Any>
    ): CompletableFuture<Product?> {
        val id = representation["id"]?.toString()
        // use dataloader to resolve Product by id
        return environment.getDataLoader<String, Product?>("ProductDataLoader").load(id)
    }
}
```

2. Define your DataLoader
```kotlin
@Service
class ProductDataLoader : KotlinDataLoader<ID, User> {
    override val dataLoaderName = "ProductDataLoader"
    override fun getDataLoader(graphQLContext: GraphQLContext) =
        DataLoaderFactory.newDataLoader<ID, User> { ids ->
            // call your DB here
        }
}

```

**Reply by samuelAndalon** [reply]
please let me know if you need more help, i just copy-pasted what is in the documentation

**Reply by hrkfdn** [reply]
Apologies, I was looking at this but the coalescing of queries via `FederatedTypePromiseResolver` and data loaders looked a bit too magical to be true for me, as I've never worked with the data loader pattern.

So thanks again for clarifying, I got it to work now.
